man/dmplot.1
.TH "DMPLOT" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmplot \(en creates plots from observations in database
.SH SYNOPSIS
\f[B]dmplot\f[R] \-\-help
.PP
\f[B]dmplot\f[R] \-\-version
.PP
\f[B]dmplot\f[R] \-\-\f[B]database\f[R] \f[I]file\f[R]
\-\-\f[B]node\f[R] \f[I]id\f[R] \-\-\f[B]sensor\f[R] \f[I]id\f[R]
\-\-\f[B]target\f[R] \f[I]id\f[R] \-\-\f[B]response\f[R] \f[I]name\f[R]
\-\-\f[B]from\f[R] \f[I]timestamp\f[R] \-\-\f[B]to\f[R]
\f[I]timestamp\f[R] \-\-\f[B]terminal\f[R] \f[I]terminal\f[R]
[\-\-\f[B]output\f[R] \f[I]file\f[R]] [\-\-\f[B]background\f[R]
\f[I]color\f[R]] [\-\-\f[B]foreground\f[R] \f[I]color\f[R]]
[\-\-\f[B]font\f[R] \f[I]name\f[R]] [\-\-\f[B]title\f[R]
\f[I]title\f[R]] [\-\-\f[B]width\f[R] \f[I]n\f[R]]
.PP
\f[B]dmplot\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
The \f[B]dmplot\f[R] program is a front\-end to \f[I]gnuplot(1)\f[R]
that creates plots of observations read from database.
Plots are either written to file or displayed in terminal or X11 window.
.PP
Depending on the selected terminal backend, you may have to set the
environment variable \f[CR]GDFONTPATH\f[R] to the local font directory
first:
.IP
.EX
$ export GDFONTPATH=/usr/local/share/fonts/webfonts/
.EE
.PP
Plotting parameters passed via command\-line have priority over those
from configuration file.
.SH OPTIONS
.TP
\f[B]\-\-background\f[R], \f[B]\-G\f[R] \f[I]color\f[R]
Optional background color of the plot (for example, \f[CR]#ff0000\f[R]
or \f[CR]red\f[R]).
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
File path to the configuration file.
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
File path to the SQLite observation database.
.TP
\f[B]\-\-font\f[R], \f[B]\-A\f[R] \f[I]name\f[R]
Optional font name (for example, \f[CR]Open\f[R] \f[CR]Sans\f[R],
\f[CR]arial.ttf\f[R], or \f[CR]monospace\f[R]).
.TP
\f[B]\-\-foreground\f[R], \f[B]\-P\f[R] \f[I]color\f[R]
Optional foreground color of the plot (for example, \f[CR]#ffffff\f[R]
or \f[CR]white\f[R]).
.TP
\f[B]\-\-from\f[R], \f[B]\-B\f[R] \f[I]timestamp\f[R]
Start of time range in ISO 8601.
.TP
\f[B]\-\-height\f[R], \f[B]\-H\f[R] \f[I]n\f[R]
Optional plot height in pixels.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of program instance and configuration (default is
\f[CR]dmplot\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-output\f[R], \f[B]\-o\f[R] \f[I]file\f[R]
File path of the output image.
Required for the terminals \f[CR]gif\f[R], \f[CR]png\f[R], and
\f[CR]pngcairo\f[R].
Additional format descriptors are replaced with their values to set date
and time dynamically (\f[CR]%Y\f[R], \f[CR]%M\f[R], \f[CR]%D\f[R],
\f[CR]%h\f[R], \f[CR]%m\f[R], \f[CR]%s\f[R]).
.TP
\f[B]\-\-response\f[R], \f[B]\-R\f[R] \f[I]name\f[R]
Response name.
.TP
\f[B]\-\-sensor\f[R], \f[B]\-S\f[R] \f[I]id\f[R]
Sensor id.
.TP
\f[B]\-\-target\f[R], \f[B]\-T\f[R] \f[I]id\f[R]
Target id.
.TP
\f[B]\-\-terminal\f[R], \f[B]\-m\f[R] [ansi|ascii|gif|gpic|png|pngcairo|postscript|sixelgd|sixeltek|svg|x11]
Output terminal, either ANSI, ASCII, GIF, GPIC (troff), PNG, PNG
(libcairo), PostScript (EPS), Sixel (libgd), Sixel (bitmap), SVG, or
X11.
The terminal must be supported by \f[I]gnuplot(1)\f[R].
.TP
\f[B]\-\-title\f[R], \f[B]\-C\f[R] \f[I]title\f[R]
Plot title.
.TP
\f[B]\-\-to\f[R], \f[B]\-E\f[R] \f[I]timestamp\f[R]
End of time range in ISO 8601.
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.TP
\f[B]\-\-width\f[R], \f[B]\-W\f[R] \f[I]n\f[R]
Optional plot width in pixels.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Plot has been created.
.TP
\f[B]1\f[R]
Failure.
Plot creation failed.
.SH EXAMPLE
Create a plot of observations selected from database
\f[CR]observ.db\f[R] in PNG terminal, and write the file to
\f[CR]/tmp/plot.png\f[R]:
.IP
.EX
$ dmplot \-N dummy\-node \-S dummy\-sensor \-T dummy\-target \(rs
  \-R dummy \-B 2020 \-E 2021 \-database observ.db \(rs
  \-m pngcairo \-o /tmp/plot.png
.EE
.PP
Output the plot directly to terminal, with the configuration loaded from
file:
.IP
.EX
$ dmplot \-n dmplot \-c dmplot.conf \-m sixelgd
.EE
.PP
Output in \f[CR]sixelgd\f[R] format requires a terminal emulator with
Sixel support.


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmplot.f90
! dmplot.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmplot
    !! Creates plots from time series.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmplot'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    character(*), parameter :: APP_XLABEL = 'Time'

    type :: app_type
        !! Application settings.
        character(ID_LEN)            :: name       = APP_NAME           !! Name of instance and POSIX semaphore.
        character(FILE_PATH_LEN)     :: config     = ' '                !! Path to configuration file.
        character(FILE_PATH_LEN)     :: database   = ' '                !! Path to observation database.
        character(NODE_ID_LEN)       :: node_id    = ' '                !! Node id.
        character(SENSOR_ID_LEN)     :: sensor_id  = ' '                !! Sensor id.
        character(TARGET_ID_LEN)     :: target_id  = ' '                !! Target id.
        character(RESPONSE_NAME_LEN) :: response   = ' '                !! Response name.
        character(TIME_LEN)          :: from       = ' '                !! Start of time range (ISO 8601).
        character(TIME_LEN)          :: to         = ' '                !! End of time range (ISO 8601).
        integer                      :: terminal   = PLOT_TERMINAL_NONE !! Plot terminal backend.
        character(FILE_PATH_LEN)     :: output     = ' '                !! Path of plot file.
        character(8)                 :: background = ' '                !! Background colour.
        character(8)                 :: foreground = ' '                !! Foreground colour (graph).
        character(32)                :: font       = ' '                !! Font name.
        character(80)                :: title      = ' '                !! Plot title.
        integer                      :: width      = 1000               !! Plot width.
        integer                      :: height     = 400                !! Plot height.
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments and configuration file options.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Create plot.
    rc = create_plot(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)
contains
    integer function create_graph(dps, terminal, output, background, foreground, &
                                  font, title, width, height, xlabel, ylabel) result(rc)
        !! Writes plot to file or shows X11 window.
        type(dp_type), intent(inout)        :: dps(:)     !! Data points array.
        integer,       intent(in)           :: terminal   !! Plot terminal.
        character(*),  intent(in), optional :: output     !! Output file.
        character(*),  intent(in), optional :: background !! Background colour.
        character(*),  intent(in), optional :: foreground !! Foreground colour.
        character(*),  intent(in), optional :: font       !! Plot font.
        character(*),  intent(in), optional :: title      !! Plot title.
        integer,       intent(in), optional :: width      !! Plot width.
        integer,       intent(in), optional :: height     !! Plot height.
        character(*),  intent(in), optional :: xlabel     !! X label.
        character(*),  intent(in), optional :: ylabel     !! Y label.

        type(plot_type) :: plot

        plot%terminal = terminal
        plot%persist  = (plot%terminal == PLOT_TERMINAL_X11)

        if (dm_string_is_present(output))       plot%output     = output
        if (dm_string_is_present(background))   plot%background = background
        if (dm_string_is_present(foreground))   plot%foreground = foreground
        if (dm_string_is_present(font))         plot%font       = font
        if (dm_string_is_present(title))        plot%title      = title
        if (dm_string_is_present(xlabel))       plot%xlabel     = xlabel
        if (dm_string_is_present(ylabel))       plot%ylabel     = ylabel

        if (present(width)) then
            if (width > 0) plot%width = width
        end if

        if (present(height)) then
            if (height > 0) plot%height = height
        end if

        rc = dm_plot_lines(plot, dps)
    end function create_graph

    integer function create_plot(app) result(rc)
        type(app_type), intent(inout) :: app !! App type.

        character(:),  allocatable :: path
        type(dp_type), allocatable :: dps(:)

        plot_block: block
            ! Read data points from database.
            rc = read_data_points(dps, app%database, app%node_id, app%sensor_id, app%target_id, &
                                  app%response, app%from, app%to)
            if (dm_is_error(rc)) exit plot_block

            ! Parse output path for format descriptors.
            path = dm_time_parse_string(app%output)

            ! Create plot.
            rc = create_graph(dps, app%terminal, path, app%background, app%foreground, app%font, &
                              app%title, app%width, app%height, APP_XLABEL, app%response)
        end block plot_block

        if (rc == E_DB_NO_ROWS) then
            call dm_error_out(rc, 'no observations found in database ' // app%database)
        else if (dm_is_error(rc)) then
            call dm_error_out(rc)
        end if
    end function create_plot

    integer function read_data_points(dps, database, node, sensor, target, response, from, to) result(rc)
        !! Returns data points from observations database.
        type(dp_type), allocatable, intent(out) :: dps(:)   !! Returned data points from database.
        character(*),               intent(in)  :: database !! Path to database.
        character(*),               intent(in)  :: node     !! Node id.
        character(*),               intent(in)  :: sensor   !! Sensor id.
        character(*),               intent(in)  :: target   !! Target id.
        character(*),               intent(in)  :: response !! Response name.
        character(*),               intent(in)  :: from     !! Start of time range.
        character(*),               intent(in)  :: to       !! End of time range.

        type(db_type) :: db

        db_block: block
            rc = dm_db_open(db, database, read_only=.true.)
            if (dm_is_error(rc)) exit db_block
            rc = dm_db_select(db, dps, node, sensor, target, response, from, to, error=E_NONE)
        end block db_block

        call dm_db_close(db)
    end function read_data_points

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from file.
        type(app_type), intent(out) :: app !! App type.

        character(PLOT_TERMINAL_NAME_LEN) :: terminal_name
        type(arg_parser_class)            :: parser

        call parser%add('name',       short='n', type=ARG_TYPE_ID)                     ! -n, --name <string>
        call parser%add('config',     short='c', type=ARG_TYPE_FILE)                   ! -c, --config <path>
        call parser%add('database',   short='d', type=ARG_TYPE_DATABASE, exist=.true.) ! -d, --database <file>
        call parser%add('node',       short='N', type=ARG_TYPE_ID)                     ! -N, --node <id>
        call parser%add('sensor',     short='S', type=ARG_TYPE_ID)                     ! -S, --sensor <id>
        call parser%add('target',     short='T', type=ARG_TYPE_ID)                     ! -T, --target <id>
        call parser%add('response',   short='R', type=ARG_TYPE_ID,     max_len=RESPONSE_NAME_LEN) ! -R, --response <name>
        call parser%add('from',       short='B', type=ARG_TYPE_TIME)                   ! -B, --from <timestamp>
        call parser%add('to',         short='E', type=ARG_TYPE_TIME)                   ! -E, --to <timestamp>
        call parser%add('terminal',   short='m', type=ARG_TYPE_STRING, max_len=PLOT_TERMINAL_NAME_LEN) ! -m, --terminal <name>
        call parser%add('output',     short='o', type=ARG_TYPE_FILE)                   ! -o, --output <path>
        call parser%add('background', short='G', type=ARG_TYPE_STRING)                 ! -G, --background <color>
        call parser%add('foreground', short='P', type=ARG_TYPE_STRING)                 ! -P, --foreground <color>
        call parser%add('font',       short='A', type=ARG_TYPE_STRING)                 ! -A, --font <name>
        call parser%add('title',      short='C', type=ARG_TYPE_STRING)                 ! -C, --title <title>
        call parser%add('width',      short='W', type=ARG_TYPE_INTEGER)                ! -W, --width <n>
        call parser%add('height',     short='H', type=ARG_TYPE_INTEGER)                ! -H, --height <n>

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Read all other options.
        call parser%get('database',   app%database)
        call parser%get('node',       app%node_id)
        call parser%get('sensor',     app%sensor_id)
        call parser%get('target',     app%target_id)
        call parser%get('response',   app%response)
        call parser%get('from',       app%from)
        call parser%get('to',         app%to)
        call parser%get('terminal',   terminal_name)
        call parser%get('output',     app%output)
        call parser%get('background', app%background)
        call parser%get('foreground', app%foreground)
        call parser%get('font',       app%font)
        call parser%get('title',      app%title)
        call parser%get('width',      app%width)
        call parser%get('height',     app%height)

        app%terminal = dm_plot_terminal_from_name(terminal_name)

        select case (app%terminal)
            case (PLOT_TERMINAL_ANSI, PLOT_TERMINAL_SIXELGD, PLOT_TERMINAL_SIXELTEK, PLOT_TERMINAL_X11)
                app%output = ' ' ! Ignore output file path.
        end select

        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads app configuration from (Lua) file.
        type(app_type), intent(inout) :: app !! App type.

        character(PLOT_TERMINAL_NAME_LEN) :: terminal_name
        type(config_class)                :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('background', app%background)
            call config%get('database',   app%database)
            call config%get('font',       app%font)
            call config%get('foreground', app%foreground)
            call config%get('terminal',   terminal_name)
            call config%get('from',       app%from)
            call config%get('height',     app%height)
            call config%get('node',       app%node_id)
            call config%get('output',     app%output)
            call config%get('response',   app%response)
            call config%get('sensor',     app%sensor_id)
            call config%get('target',     app%target_id)
            call config%get('title',      app%title)
            call config%get('to',         app%to)
            call config%get('width',      app%width)

            app%terminal = dm_plot_terminal_from_name(terminal_name)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_file_exists(app%database)) then
            call dm_error_out(rc, 'database ' // trim(app%database) // ' not found')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid or missing node id')
            return
        end if

        if (.not. dm_id_is_valid(app%sensor_id)) then
            call dm_error_out(rc, 'invalid or missing sensor id')
            return
        end if

        if (.not. dm_id_is_valid(app%target_id)) then
            call dm_error_out(rc, 'invalid or missing target id')
            return
        end if

        if (.not. dm_id_is_valid(app%response)) then
            call dm_error_out(rc, 'invalid or missing response name')
            return
        end if

        if (.not. dm_time_is_valid(app%from)) then
            call dm_error_out(rc, 'invalid or missing timestamp from')
            return
        end if

        if (.not. dm_time_is_valid(app%to)) then
            call dm_error_out(rc, 'invalid or missing timestamp to')
            return
        end if

        if (.not. dm_plot_terminal_is_valid(app%terminal)) then
            call dm_error_out(rc, 'invalid or missing plot terminal')
            return
        end if

        if (app%width < 0) then
            call dm_error_out(rc, 'invalid plot width')
            return
        end if

        if (app%height < 0) then
            call dm_error_out(rc, 'invalid plot height')
            return
        end if

        select case (app%terminal)
            case (PLOT_TERMINAL_GIF,        &
                  PLOT_TERMINAL_GPIC,       &
                  PLOT_TERMINAL_PNG,        &
                  PLOT_TERMINAL_PNGCAIRO,   &
                  PLOT_TERMINAL_POSTSCRIPT, &
                  PLOT_TERMINAL_SVG)
                if (.not. dm_string_has(app%output)) then
                    call dm_error_out(rc, 'missing output path')
                    return
                end if
        end select

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        logical :: found

        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a, 2(1x, a))', dm_plot_version(.true., found), dm_lua_version(.true.), dm_db_version(.true.)
        if (.not. found) call dm_error_out(E_NOT_FOUND, 'Gnuplot binary not found')
    end subroutine version_callback
end program dmplot
